### PR TITLE
Save restart (snapshot) information during filter run

### LINKED ISSUE #176: Filter should allow saving restart
Not hard, just a bit more book-keeping

### PR DESCRIPTION
**Pull request by richfitz**
A bit of a terminology change here, but I think this is better and I might update in mcstate too.

Fixes #176

### CODE REVIEW

File: inst/include/dust/dust.hpp [hunk 1]
```diff
@@ -520,17 +532,22 @@ class Dust {
       }
 
       // We could move this below if wanted but we'd have to rewrite
-      // the re-sort algorithm.
-      if (save_history) {
-        state(filter_state_.history_value_iterator());
+      // the re-sort algorithm; that would be worth doing I think
```
Comment (johnlees) [inline inst/include/dust/dust.hpp:535]: Can you expand on this/add an issue? Does 'below' mean after running `resample`?

File: tests/testthat/test-filter.R [hunk 2]
```diff
@@ -126,3 +128,63 @@ test_that("scale log weights copes with NaN", {
...
+  dat <- example_filter()
+
+  np <- 10
+  mod <- dat$model$new(list(), 0, np, seed = 10L)
+  mod$set_data(dat$dat_dust)
+
+  ## TODO: handle better doubles here using as_integer
+  step_snapshot <- c(100L, 200L, 400L)
+
+  ## We can perform the entire particle filter manually with the C
```
Comment (johnlees) [inline tests/testthat/test-filter.R:145]:
```suggestion
  ## We can perform the entire particle filter manually with the C++
```

File: tests/testthat/test-filter.R [hunk 3]
```diff
@@ -126,3 +128,63 @@ test_that("scale log weights copes with NaN", {
   expect_equal(res$weights[[3]], 0)
   expect_equal(res, scale_log_weights(w))
 })
+
+
+## This is badly wrong: we've written into some incorrect memory and
+## we never aligned correctly with the steps! Baby steps.
```
Comment (johnlees) [inline tests/testthat/test-filter.R:134]: But fixed now? Or where is this bad write happening?

**Review by johnlees — CHANGES_REQUESTED** [review]

File: inst/include/dust/types.hpp [hunk 4]
```diff
@@ -135,6 +133,69 @@ class filter_state {
...
+    n_particles_ = n_particles;
+    n_steps_ = steps.size();
+    offset_ = 0;
+    steps_ = steps;
+    state_.resize(n_state_ * n_particles_ * n_steps_);
+  }
+
+  bool collect(size_t step) {
+    return offset_ < n_steps_ && steps_[offset_] == step;
+  }
```
[the remark this answers is not in the document]
Comment (richfitz) [reply inst/include/dust/types.hpp:156]: It's testing if this is a time that we should collect a shapshot for, I'll add a comment and possibly rename the method to make it clearer

File: tests/testthat/test-filter.R (hunk 3 above)
Comment (richfitz) [reply tests/testthat/test-filter.R:134]: oops, my tests are always littered with notes to myself like this when refactoring...

File: inst/include/dust/dust.hpp (hunk 1 above)
Comment (richfitz) [reply inst/include/dust/dust.hpp:535]: yeah, exactly. Atm we collect the state, then resort - when reconstructing the tree we rely on that order. But we could do both steps at once which reduces the amount of logic in the function but requires rewriting some very painful code

**Review by johnlees — APPROVED** [review]

**Pull request merged**